MaslowDue/system.cpp

```diff
@@ -395,113 +401,219 @@ uint8_t system_check_travel_limits(float *target)
-//      *yTarget = (*yTarget < -halfHeight) ? -halfHeight : (*yTarget > halfHeight) ? halfHeight : *yTarget;
-  
-  }
-
-// calculate left and right (LEFT_MOTOR/RIGHT_MOTOR) chain lengths from X-Y cartesian coordinates  (in mm)
-// target is an absolute position in the frame
-  void triangularInverse(float xTarget,float yTarget, float* aChainLength, float* bChainLength)
-  {
-      //Confirm that the coordinates are on the table
-      verifyValidTarget(&xTarget, &yTarget);
-
       // scale target (absolute position) by any correction factor
+      // Use double math internally for faster computation.
       double xxx = (double)xTarget * (double)settings.XcorrScaling;
       double yyy = (double)yTarget * (double)settings.YcorrScaling;
-  
+
       //Calculate motor axes length to the bit
       double Motor1Distance = sqrt(pow((double)(-1*_xCordOfMotor) - (double)(xxx),2)+pow((double)(_yCordOfMotor) - (double(yyy)),2));
       double Motor2Distance = sqrt(pow((double)   (_xCordOfMotor) - (double)(xxx),2)+pow((double)(_yCordOfMotor) - (double)(yyy),2));
 
-      *aChainLength = Motor1Distance;
-      *bChainLength = Motor2Distance;
-      return;
+      //Set up variables
+      double Chain1Angle = 0, Chain2Angle = 0;
+      double Chain1AroundSprocket = 0, Chain2AroundSprocket = 0;
+
+      //Calculate the chain angles from horizontal, based on if the chain connects to the sled from the top or bottom of the sprocket
+      if(settings.chainOverSprocket == 1){
+        Chain1Angle = asin((_yCordOfMotor - yTarget)/Motor1Distance) + asin(_sprocketRadius/Motor1Distance);
+        Chain2Angle = asin((_yCordOfMotor - yTarget)/Motor2Distance) + asin(_sprocketRadius/Motor2Distance);
+        Chain1AroundSprocket = _sprocketRadius * Chain1Angle;
+        Chain2AroundSprocket = _sprocketRadius * Chain2Angle;
+      } else {
+        Chain1Angle = asin((_yCordOfMotor - yTarget)/Motor1Distance) - asin(_sprocketRadius/Motor1Distance);
+        Chain2Angle = asin((_yCordOfMotor - yTarget)/Motor2Distance) - asin(_sprocketRadius/Motor2Distance);
+        Chain1AroundSprocket = _sprocketRadius * (3.14159 - Chain1Angle);
+        Chain2AroundSprocket = _sprocketRadius * (3.14159 - Chain2Angle);
+      }
+
+      //Calculate the straight chain length from the sprocket to the bit
+      double srsqrd = pow(_sprocketRadius,2);
+      double Chain1Straight = sqrt(pow(Motor1Distance,2)-srsqrd);
+      double Chain2Straight = sqrt(pow(Motor2Distance,2)-srsqrd);
+
+      //Correct the straight chain lengths to account for chain sag
+      double csc = (settings.chainSagCorrection / 1000000000000);
+      Chain1Straight *= (1 + (csc * pow(cos(Chain1Angle),2) * pow(Chain1Straight,2) * pow((tan(Chain2Angle) * cos(Chain1Angle)) + sin(Chain1Angle),2)));
+      Chain2Straight *= (1 + (csc * pow(cos(Chain2Angle),2) * pow(Chain2Straight,2) * pow((tan(Chain1Angle) * cos(Chain2Angle)) + sin(Chain2Angle),2)));
```

Comment:
So this is the "old" chain sag calculation.

Within the Maslow Firmware, the new calculation doesn't use a "sag" correction.  But instead attempts to calculate the catenary curve by calculating for forces on the chain.  I think this method is preferable, and should work fine here since the Mega2560 is able to keep up with it.

Sorry if I misled you by calling it "sag" I really meant incorporating the newer equations.

Reply:
Yeah, just realized today that the source code I had been linked to (and was copying from) was the original `MaslowCNC` repository. Is this the correct latest version? https://github.com/WebControlCNC/Firmware

Reply:
If the goal is to get the Holey Calibration working in the MaslowDue and I think it is, then look at the branch release/holey

https://github.com/WebControlCNC/Firmware/tree/release/holey

Reply:
Ah yes, spot on, thanks!